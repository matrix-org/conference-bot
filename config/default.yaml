# Where the homeserver's Client-Server API is located. Typically this
# is where clients would be connecting to in order to send messages.
homeserverUrl: "https://matrix-client.matrix.org"

# An access token for the bot to use. This should be the as_token from the
# appservice registration.
accessToken: "YOUR_TOKEN_HERE"

# Set this to the desired name for the bot. The account must already be
# registered and accessible as an impersonatable user ID by the accessToken.
userId: "@bot:example.org"

# Location on disk for where to store various bot information.
dataPath: "storage"

# The ID of the room to use as a management room. The bot will only
# accept commands in here.
managementRoom: "!room:example.org"

# The identity server to use to send email invites to any participants. The
# bot will accept all terms of service automatically - please ensure that it
# has legal capability to do so before running the bot in this way.
idServerDomain: "vector.im"

# The brand to supply to the identity server when performing actions which may
# result in emails or other communications to the user.
idServerBrand: "vector-im"

# This is the user ID of the moderator for all public-facing rooms the bot creates.
# Note that this user will be granted power level 100 (the highest) in every room
# and be invited.
moderatorUserId: "@moderator:example.org"

# Settings for how the bot should represent livestreams.
livestream:
  # The template for livestreams in auditorium rooms
  # Available variables:
  #  id  - The room ID/name (eg: "D.collab")
  #  sId - The room ID/name, but lowercase and only alphanumeric characters
  #        are included (eg: "dcollab")
  auditoriumUrl: "https://stream.example.org/conference/hls/{id}.m3u8"

  # The template for livestreams in talk rooms
  # Available variables:
  #  slug  - The shorthand version of the talk name (eg: "matrix-p2p")
  #  audId - The auditorium ID the talk is held in (eg: "D.collab")
  #  jitsi - The Jitsi conference name in lowercase.
  talkUrl: "https://stream.example.org/conference-talks/hls/{slug}.m3u8"

  # The template for livestreams in random rooms.
  # Available variables:
  #  jitsi - The Jitsi conference name in lowercase.
  hybridUrl: "https://stream.example.org/conference-hybrids/hls/{jitsi}.m3u8"

  # The URL template for the schedule widget.
  # Available variables:
  #  audId - The auditorium ID the talk is held in (eg: "D.collab")
  scheduleUrl: "https://example.org/?theme=$theme&room={audId}"

  # The domain of the Jitsi server to use. It will be assumed to be using
  # https://github.com/matrix-org/prosody-mod-auth-matrix-user-verification
  jitsiDomain: "jitsi.riot.im"

  # An MSC URI to use as the avatar for widgets generated by the bot.
  widgetAvatar: "mxc://example.org/mediaid"

  # Options for the /onpublish endpoint used to redirect streams from Jitsi
  # to an RTMP location. Query string arguments are:
  #  auth=secret   - See below
  #  conf=aabbcc   - The Jitsi conference name. This should be a base32 string
  #                  when using prosody-mod-auth-matrix-user-verification as above.
  # If no mapping is found, 200 OK is returned for compatibility with nginx.
  onpublish:
    # The hostname to use in the rtmpPrefix template. Some variables are supported
    # here:
    #  squishedAudId  - The auditorium ID with all non-alphanumeric characters
    #                   removed, and lowercase.
    # Note: The hostname will be resolved to an IP address before used as a hostname
    # for compatibility with nginx.
    rtmpHostnameTemplate: "{squishedAudId}.example.org"

    # The URL to redirect to, including scheme.
    # Some variables are supported here:
    #  hostname    - The hostname as generated/resolved by rtmpHostnameTemplate
    #  saltedHash  - The salted hash generated by the bot to secure the stream with.
    rtmpUrlTemplate: "rtmp://{hostname}/stream/{saltedHash}"

    # The salt to use in the hash. The hash will be "event_id.secret" where
    # "event_id" is the penta event ID.
    #
    # The value which must also be supplied as ?auth to /onpublish in order to get
    # a proper 302 redirect. When the auth mismatches a 200 status code will be
    # returned to imply there is no mapping.
    salt: "change me"

# Configuration for the webserver (used to serve widgets for the conference's
# rooms).
webserver:
  # The address to bind to. If using Docker, this should be all interfaces and
  # instead managed through port mapping.
  address: '0.0.0.0'

  # The port to bind the webserver on.
  port: 8080

  # The public URL where the bot can expect to be able to serve widgets from.
  # Note that this URL will be put into the Matrix room state.
  publicBaseUrl: "https://conferences.example.org"

  # Additional assets may be served from `${publicBaseUrl}/assets/*` which are
  # available at this path.
  additionalAssetsPath: "/data/public"

# Information about the conference
conference:
  # The ID of the conference to call this one. Data from other conferences
  # will be ignored (such as past conferences). This will also be used to create
  # the Matrix Space.
  id: "fosdem2021"

  # The name of the conference.
  name: "FOSDEM 2021"

  # The URL to the XML which is updated with conference information.
  # This is read and parsed by the bot during the early stages of
  # setting up the conference.
  pentabarfDefinition: "https://fosdem.org/2021/schedule/xml"

  # The timezone that the the bot's database is operating off of.
  timezone: "Europe/Brussels"

  # The various support rooms the bot supports. They will be created if they
  # haven't already.
  supportRooms:
    speakers: "#speakers:example.org"
    specialInterest: "#stands:example.org"
    coordinators: "#coordinators:example.org"

  # How far in the future the bot should be looking before considering something "finalized".
  # Should never be less than 5 minutes.
  lookaheadMinutes: 5

  # Connection information to get data about the conference during the conference.
  # This can be readonly. Currently the bot only supports postgresql.
  database:
    host: "localhost"
    port: 5432
    username: "db_username"
    password: "db_password"
    database: "your_db"
    sslmode: "require"

    # The table/view for the bot to access for information about people involved in the conference.
    # Must at least have the following columns:
    #  event_id         - TEXT(like) - The penta ID for the talk (event) associated with
    #                                  the person, if relevant. Null if a coordinator.
    #  person_id        - TEXT(like) - The penta ID for the person.
    #  event_role       - TEXT(like) - One of [speaker, coordinator, host]. See roles later on.
    #  name             - TEXT(like) - The full name, or otherwise useful name, for the person.
    #  email            - TEXT(like) - The preferred email address for the person.
    #  matrix_id        - TEXT(like) - If known, the Matrix User ID for the person, otherwise null.
    #  conference_room  - TEXT(like) - The relevant room. This should match the event_id or be
    #                                  the room where coordinators are assigned.
    #
    # People may be assigned multiple roles with multiple rows.
    #
    # Roles:
    #  speaker      - A speaker for event_id. This person will get moderator in their talk's
    #                 room, and invited to the auditorium backstage room. They will be required
    #                 to check in before their talk starts.
    #
    #  host         - Someone who is moderating the talk itself in collaboration with the speakers
    #                 for the event_id. One host must check in before the talk starts, otherwise
    #                 the issue will be raised to the management room and coordinators will be
    #                 asked to take over. Hosts will get moderator in their talk's room, and be
    #                 invited to the public auditorium room, backstage room, and talk room.
    #
    #  coordinator  - These are typically people responsible for scheduling the auditorium's
    #                 talks. They'll get moderator in the auditorium room, backstage room, and all
    #                 talk rooms for their auditorium, though will only be invited to the auditorium
    #                 and backstage room (and the talk rooms on-demand if needed to fill in for a
    #                 missing host).
    #
    # The bot will ignore unknown events, rooms, and roles.
    #
    # CAUTION: Although the bot uses parameterized queries, it is unable to use the table name as
    # a parameter. As such, this particular config value is vulnerable to SQL injection. Seeing as
    # how you (the bot's admin) are the one entering it: don't do that to yourself.
    tblPeople: "view_matrix_bot_export_people"

    # The table/view for the bot to access for information about the conference schedule.
    # Must at least have the following columns:
    #  event_id            - TEXT(like) - The penta ID for the talk (event).
    #  conference_room     - TEXT(like) - The relevant room. This should match the event_id.
    #  start_datetime      - TIMESTAMP WITHOUT TIME ZONE - The start time of the talk. The timezone
    #                        is determined by the bot using the timezone elsewhere in this config.
    #  duration            - INTERVAL   - How long the talk is meant to last (start to end).
    #  presentation_length - INTERVAL   - The portion of `duration` where the speaker is presenting.
    #                                     The remainder is assumed to be Q&A.
    #  prerecorded         - BOOLEAN    - True if the presentation is prerecorded. When false, the
    #                                     bot will assume that presentation_length is zero and switch
    #                                     directly into "hallway" mode at start_datetime.
    #
    # The bot will ignore unknown talks. Any if the start_datetime is NULL then the bot will ignore
    # that record. If the INTERVAL fields are NULL, the bot will assume zero.
    #
    # Note: the bot is capable of running both tblSchedule and tblPeople off the same table/view.
    # It will use DISTINCT where it needs to in order to get accurate information.
    #
    # CAUTION: Although the bot uses parameterized queries, it is unable to use the table name as
    # a parameter. As such, this particular config value is vulnerable to SQL injection. Seeing as
    # how you (the bot's admin) are the one entering it: don't do that to yourself.
    tblSchedule: "view_matrix_bot_export_schedule"

    # The time in seconds added to the presentation_length to account for preroll material such
    # as sponsor segments. If this is built into the table already then set this to zero.
    scheduleBufferSeconds: 30

  # Various prefixes used by the bot when parsing information.
  prefixes:
    # The prefixes for the rooms listed in the pentabarf definition which
    # describe the rooms that are 'auditorium' rooms. The prefixes will be
    # trimmed when naming the room (unless an override is present).
    auditoriumRooms: ["D."]

    # The prefixes for the rooms listed in the pentabarf definition which
    # describe the rooms that are 'special interest' rooms. The prefixes will
    # be trimmed when naming the room (unless an override is present).
    interestRooms: ["S.", "B."]

    # Simple mapping of room IDs to chosen names. The names should not contain
    # spaces or any complex characters. This is typically used to resolve naming
    # conflicts after trimming the prefix.
    nameOverrides:
      #"D.example": "example-room"

    # The prefix to apply to all generated room aliases.
    aliases: ""

    # The alias suffixes to apply for certain prefixes of rooms
    suffixes:
      #"S.": "-stand"

    # The display name suffixes to apply for certain prefixes of rooms
    displayNameSuffixes:
      #"S.": " stand"

  # A mapping of pre-existing rooms to re-use as interest rooms.
  existingInterestRooms:
    #"I.infodesk": "#infodesk:example.org"

  # The subspaces to create under the main conference space
  subspaces:
    #stands:
    #  displayName: Stands
    #  alias: stands
    #  # The prefixes of rooms which belong in the subspace
    #  prefixes: ["S."]

# Options related to the IRC bridge.
ircBridge:
  # The nick to use when taking ownership of channels
  botNick: 'confbot'
  botPassword: 'a_secret_password'

  # The server name of the IRCd
  serverName: 'localhost'

  # The port of the IRCd
  port: 6667

  # The userId of the irc bridge
  botUserId: '@appservice-irc:example.com'

  # The allowed channel prefix of the bot
  channelPrefix: '#conference-'
