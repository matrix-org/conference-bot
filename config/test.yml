# We don't set any config for tests currently, but node-config complains if we don't have this file.

