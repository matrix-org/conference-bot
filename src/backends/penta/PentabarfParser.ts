/*
Copyright 2020, 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as parser from 'fast-xml-parser';
import { IAuditorium, IConference, IInterestRoom, IPerson, ITalk, Role } from "../../models/schedule";
import * as moment from "moment";
import { RoomKind } from "../../models/room_kinds";
import { IPrefixConfig } from "../../config";

function arrayLike<T>(val: T | T[]): T[] {
    if (Array.isArray(val)) return val;
    return [val];
}

function simpleTimeParse(str: string): { hours: number, minutes: number } {
    const parts = str.split(':');
    return {hours: Number(parts[0]), minutes: Number(parts[1])};
}

export function deprefix(id: string, prefixConfig: IPrefixConfig): {kind: RoomKind, name: string} {
    const override = prefixConfig.nameOverrides[id];

    const auditoriumPrefix = prefixConfig.auditoriumRooms.find(p => id.startsWith(p));
    if (auditoriumPrefix) {
        return {kind: RoomKind.Auditorium, name: override || id.substring(auditoriumPrefix.length)};
    }

    const interestPrefix = prefixConfig.interestRooms.find(p => id.startsWith(p));
    if (interestPrefix) {
        return {kind: RoomKind.SpecialInterest, name: override || id.substring(interestPrefix.length)};
    }

    return {kind: RoomKind.SpecialInterest, name: override || id};
}

export interface IPentabarfEvent {
    attr: {
        "@_id": string; // number
    };
    start: string;
    duration: string;
    room: string;
    slug: string;
    title: string;
    subtitle: string;
    track: string;
    type: "devroom";
    language: string;
    abstract: string;
    description: string;
    persons: {
        person: {
            attr: {
                "@_id": string; // number
            };
            "#text": string;
        }[];
    };
    attachments: unknown; // TODO
    links: {
        link: {
            attr: {
                "@_href": string;
            };
            "#text": string;
        }[];
    };
}

export interface IPentabarfSchedule {
    schedule: {
        conference: {
            title: string;
            subtitle: string;
            venue: string;
            city: string;
            start: string;
            end: string;
            days: number;
            day_change: string;
            timeslot_duration: string;
        };
        day: {
            attr: {
                "@_index": string; // number
                "@_date": string;
            };
            room: {
                attr: {
                    "@_name": string;
                };
                event: IPentabarfEvent[];
            }[];
        }[];
    };
}

export class PentabarfParser {
    private readonly parsed: IPentabarfSchedule;

    public readonly conference: IConference;
    public readonly auditoriums: IAuditorium[];
    public readonly talks: ITalk[];
    public readonly speakers: IPerson[];
    public readonly interestRooms: IInterestRoom[];

    constructor(rawXml: string, prefixConfig: IPrefixConfig) {
        this.parsed = parser.parse(rawXml, {
            attrNodeName: "attr",
            textNodeName: "#text",
            ignoreAttributes: false,
        });

        this.auditoriums = [];
        this.talks = [];
        this.speakers = [];
        this.interestRooms = [];
        this.conference = {
            title: this.parsed.schedule?.conference?.title,
            auditoriums: this.auditoriums,
            interestRooms: this.interestRooms,
        };

        for (const day of arrayLike(this.parsed.schedule?.day)) {
            if (!day) continue;

            const dateTs = moment.utc(day.attr?.["@_date"], "YYYY-MM-DD").valueOf();
            for (const pRoom of arrayLike(day.room)) {
                if (!pRoom) continue;

                const metadata = deprefix(pRoom.attr?.["@_name"] || "org.matrix.confbot.unknown", prefixConfig);
                if (metadata.kind === RoomKind.SpecialInterest) {
                    let spiRoom: IInterestRoom = {
                        id: pRoom.attr?.["@_name"],
                        name: metadata.name,
                        kind: metadata.kind,
                    };
                    const existingSpi = this.interestRooms.find(r => r.id === spiRoom.id);
                    if (!existingSpi) {
                        this.interestRooms.push(spiRoom);
                    }
                    continue;
                }
                if (metadata.kind !== RoomKind.Auditorium) continue;
                let auditorium: IAuditorium = {
                    id: pRoom.attr?.["@_name"],
                    slug: metadata.name,
                    name: metadata.name,
                    kind: metadata.kind,
                    talks: new Map(),
                };
                const existingAuditorium = this.auditoriums.find(r => r.id === auditorium.id);
                if (existingAuditorium) {
                    auditorium = existingAuditorium;
                } else {
                    this.auditoriums.push(auditorium);
                }

                const qaEnabled = prefixConfig.qaAuditoriumRooms.find(p => auditorium.id.startsWith(p)) !== undefined;

                for (const pEvent of arrayLike(pRoom.event)) {
                    if (!pEvent) continue;

                    const parsedStartTime = simpleTimeParse(pEvent.start);
                    const parsedDuration = simpleTimeParse(pEvent.duration);
                    const startTime = moment(dateTs).add(parsedStartTime.hours, 'hours').add(parsedStartTime.minutes, 'minutes');
                    const endTime = moment(startTime).add(parsedDuration.hours, 'hours').add(parsedDuration.minutes, 'minutes');
                    let talk: ITalk = {
                        id: pEvent.attr?.["@_id"],
                        dateTs: dateTs,
                        startTime: startTime.valueOf(),
                        endTime: endTime.valueOf(),
                        slug: pEvent.slug,
                        title: pEvent.title,
                        subtitle: pEvent.subtitle,
                        track: pEvent.track,
                        speakers: [],
                        prerecorded: true,
                        auditoriumId: auditorium.id,
                        livestream_endTime: 0,
                        // 0 is populated later by consulting the PentaDb.
                        qa_startTime: qaEnabled ? 0 : null,
                    };
                    const existingTalk = this.talks.find(e => e.id === talk.id);
                    if (existingTalk) {
                        talk = existingTalk;
                    } else {
                        this.talks.push(talk);
                    }

                    if (auditorium.talks.has(talk.id)) {
                        throw new Error(`Auditorium ${auditorium.id}: Talk ${talk.id}: this talk already exists and is defined a second time.`);
                    } else {
                        auditorium.talks.set(talk.id, talk);
                    }

                    for (const pPerson of arrayLike(pEvent.persons?.person)) {
                        if (!pPerson) continue;

                        let person: IPerson = {
                            id: pPerson.attr?.["@_id"],
                            name: pPerson["#text"],
                            email: "",
                            matrix_id: "",
                            role: Role.Speaker,
                        };
                        const existingPerson = this.speakers.find(s => s.id === person.id);
                        if (existingPerson) {
                            person = existingPerson;
                        } else {
                            this.speakers.push(person);
                        }

                        talk.speakers.push(person);
                    }
                }
            }
        }
    }
}
