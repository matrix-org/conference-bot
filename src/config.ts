/*
Copyright 2020, 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as config from "config";
import { MatrixClient } from "matrix-bot-sdk";
import { Conference } from "./Conference";
import { IRCBridge, IRCBridgeOpts } from "./IRCBridge";
import { Scheduler } from "./Scheduler";
import { CheckInMap } from "./CheckInMap";

interface IConfig {
    homeserverUrl: string;
    accessToken: string;
    userId: string;
    dataPath: string;
    managementRoom: string;
    idServerDomain: string;
    idServerBrand: string;
    moderatorUserId: string;
    livestream: {
        auditoriumUrl: string;
        talkUrl: string;
        hybridUrl: string;
        scheduleUrl: string;
        jitsiDomain: string;
        widgetAvatar: string;
        onpublish: {
            rtmpHostnameTemplate: string;
            rtmpUrlTemplate: string;
            salt: string;
        };
    };
    webserver: {
        address: string;
        port: number;
        publicBaseUrl: string;
        additionalAssetsPath: string;
    };
    conference: {
        id: string;
        name: string;
        pentabarfDefinition: string;
        timezone: string;
        lookaheadMinutes: number;
        supportRooms: {
            speakers: string;
            coordinators: string;
            specialInterest: string;
        };
        prefixes: {
            auditoriumRooms: string[];
            interestRooms: string[];
            aliases: string;
            suffixes: {
                [prefix: string]: string;
            };
            displayNameSuffixes: {
                [prefix: string]: string;
            };
            nameOverrides: {
                [auditoriumId: string]: string;
            };
        };
        database: {
            host: string;
            port: number;
            username: string;
            password: string;
            database: string;
            sslmode: string;
            tblPeople: string;
            tblSchedule: string;
            scheduleBufferSeconds: number;
        };
    };
    ircBridge: IRCBridgeOpts;

    RUNTIME: {
        client: MatrixClient;
        conference: Conference;
        scheduler: Scheduler;
        ircBridge: IRCBridge;
        checkins: CheckInMap;
    };
}

export default <IConfig>config;
