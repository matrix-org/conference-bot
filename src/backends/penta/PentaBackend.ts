import config, { IPentaScheduleBackendConfig } from "../../config";
import { IConference, ITalk, IAuditorium, IInterestRoom, IPerson } from "../../models/schedule";
import { AuditoriumId, InterestId, IScheduleBackend, TalkId } from "../IScheduleBackend";
import { PentaDb } from "./db/PentaDb";
import { PentabarfParser } from "./PentabarfParser";
import * as fetch from "node-fetch";


export class PentaBackend implements IScheduleBackend {
    constructor(private cfg: IPentaScheduleBackendConfig, parser: PentabarfParser, public db: PentaDb) {
        this.updateFromParser(parser);
    }

    /**
     * We need an async version of the constructor.
     * Must be called right after construction.
     */
    async init() {
        await this.hydrateFromDatabase();
    }

    private updateFromParser(parser: PentabarfParser): void {
        const conference = parser.conference;
        const talks = new Map();
        const auditoriums = new Map();
        const interestRooms = new Map();

        for (let auditorium of parser.auditoriums) {
            if (auditoriums.has(auditorium.id)) {
                throw `Conflict in auditorium ID «${auditorium.id}»!`;
            }
            auditoriums.set(auditorium.id, auditorium);

            for (let talk of auditorium.talks.values()) {
                if (talks.has(talk.id)) {
                    const conflictingTalk = talks.get(talk.id);
                    throw `Talk ID ${talk.id} is not unique — occupied by both «${talk.title}» and «${conflictingTalk.title}»!`;
                }
                talks.set(talk.id, talk);
            }
        }

        for (let interest of parser.interestRooms) {
            if (interestRooms.has(interest.id)) {
                throw `Conflict in interest ID «${interest.id}»!`;
            }
        }

        // Update all at the end, to prevent non-atomic updates in the case of a failure.
        this.conference = conference;
        this.talks = talks;
        this.auditoriums = auditoriums;
        this.interestRooms = interestRooms;
    }

    private async hydrateFromDatabase(): Promise<void> {
        for (let talk of this.talks.values()) {
            this.hydrateTalk(talk);

            for (let person of talk.speakers) {
                this.hydratePerson(person);
            }
        }

        // TODO do we need to hydrate any other objects?
    }

    private async hydrateTalk(talk: ITalk): Promise<void> {
        const dbTalk = await this.db.getTalk(talk.id);
        if (dbTalk === null) return;

        if (talk.qa_startTime !== null) {
            // hydrate Q&A time if enabled
            talk.qa_startTime = dbTalk.qa_start_datetime;
        }
        talk.livestream_endTime = dbTalk.livestream_end_datetime;
    }

    private async hydratePerson(person: IPerson): Promise<void> {
        const dbPeople = await this.db.findPeopleWithId(person.id);
        if (dbPeople.length == 0) return;

        if (dbPeople.length > 1) {
            throw new Error(`Person ID '${person.id}' has ${dbPeople.length} different people associated with it!`);
        }

        const dbPerson = dbPeople[0];
        person.matrix_id = dbPerson.matrix_id;
        person.email = dbPerson.email;
    }

    wasLoadedFromCache(): boolean {
        // Penta backend doesn't support using a cache.
        return false;
    }

    static async new(cfg: IPentaScheduleBackendConfig): Promise<PentaBackend> {
        const xml = await fetch(cfg.scheduleDefinition).then(r => r.text());
        const parsed = new PentabarfParser(xml, config.conference.prefixes);
        const db = new PentaDb(cfg.database);
        await db.connect();
        const backend = new PentaBackend(cfg, parsed, db);
        await backend.init();
        return backend;
    }

    refresh(): Promise<void> {
        throw new Error("refresh() not implemented for Penta backend.");
    }

    conference: IConference;
    talks: Map<TalkId, ITalk>;
    auditoriums: Map<AuditoriumId, IAuditorium>;
    interestRooms: Map<InterestId, IInterestRoom>;
}