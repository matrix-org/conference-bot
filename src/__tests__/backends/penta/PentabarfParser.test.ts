import { PentabarfParser } from "../../../backends/penta/PentabarfParser";
import { test, expect } from "@jest/globals";
import { IPrefixConfig } from "../../../config";

const fs = require("fs");
const path = require("path");

const prefixConfig: IPrefixConfig = {
    // Unused here.
    aliases: "", displayNameSuffixes: {}, suffixes: {},

    auditoriumRooms: [
        "A.",
    ],
    interestRooms: [
        "X."
    ],
    nameOverrides: {
        "A.special": "special-room",
    },
};

/*
 * A somewhat vague test that just loads in a basic file and checks a few things, comparing against the snapshots in
 * __snapshots__.
 */
test('parsing pentabarf XML: overview', () => {
    const xml = fs.readFileSync(path.join(__dirname, "pentabarf01_overview.xml"), 'utf8');
    const p = new PentabarfParser(xml, prefixConfig);

    // TODO the auditorium id and slug look dodgy and not id-like or slug-like...
    expect(p.auditoriums).toMatchSnapshot("auditoriums");

    expect(p.conference).toMatchSnapshot("conference");

    expect(p.interestRooms).toMatchSnapshot("interestRooms");

    // NOTE Speakers don't have contact information: that's filled in by the PentaDb in the PentaBackend afterwards.
    expect(p.speakers).toMatchSnapshot("speakers");

    // NOTE I don't like that qa_startTime and livestream_endTime are 0 â€” they are updated in the PentaBackend
    //      using the PentaDb afterwards.
    expect(p.talks).toMatchSnapshot("talks");
});
