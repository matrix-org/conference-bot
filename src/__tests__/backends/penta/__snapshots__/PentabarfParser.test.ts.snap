// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsing pentabarf XML: overview: auditoriums 1`] = `
[
  {
    "id": "A.01 (Someroom)",
    "isPhysical": false,
    "kind": "auditorium",
    "name": "01 (Someroom)",
    "slug": "01_someroom_",
    "talks": Map {
      "E001" => {
        "auditoriumId": "A.01 (Someroom)",
        "dateTs": 1675468800000,
        "endTime": 1675502100000,
        "id": "E001",
        "livestream_endTime": 0,
        "prerecorded": true,
        "qa_startTime": null,
        "slug": "testcon_opening_remarks",
        "speakers": [
          {
            "email": "",
            "id": "9012",
            "matrix_id": "",
            "name": "John Teststone",
            "role": "speaker",
          },
          {
            "email": "",
            "id": "90367",
            "matrix_id": "",
            "name": "Evå 完牲 Exampleson",
            "role": "speaker",
          },
        ],
        "startTime": 1675501200000,
        "subtitle": "",
        "title": "Testcon01: Opening Remarks",
        "track": "Track01",
      },
    },
  },
]
`;

exports[`parsing pentabarf XML: overview: conference 1`] = `
{
  "auditoriums": [
    {
      "id": "A.01 (Someroom)",
      "isPhysical": false,
      "kind": "auditorium",
      "name": "01 (Someroom)",
      "slug": "01_someroom_",
      "talks": Map {
        "E001" => {
          "auditoriumId": "A.01 (Someroom)",
          "dateTs": 1675468800000,
          "endTime": 1675502100000,
          "id": "E001",
          "livestream_endTime": 0,
          "prerecorded": true,
          "qa_startTime": null,
          "slug": "testcon_opening_remarks",
          "speakers": [
            {
              "email": "",
              "id": "9012",
              "matrix_id": "",
              "name": "John Teststone",
              "role": "speaker",
            },
            {
              "email": "",
              "id": "90367",
              "matrix_id": "",
              "name": "Evå 完牲 Exampleson",
              "role": "speaker",
            },
          ],
          "startTime": 1675501200000,
          "subtitle": "",
          "title": "Testcon01: Opening Remarks",
          "track": "Track01",
        },
      },
    },
  ],
  "interestRooms": [],
  "title": "Testcon01",
}
`;

exports[`parsing pentabarf XML: overview: interestRooms 1`] = `[]`;

exports[`parsing pentabarf XML: overview: speakers 1`] = `
[
  {
    "email": "",
    "id": "9012",
    "matrix_id": "",
    "name": "John Teststone",
    "role": "speaker",
  },
  {
    "email": "",
    "id": "90367",
    "matrix_id": "",
    "name": "Evå 完牲 Exampleson",
    "role": "speaker",
  },
]
`;

exports[`parsing pentabarf XML: overview: talks 1`] = `
[
  {
    "auditoriumId": "A.01 (Someroom)",
    "dateTs": 1675468800000,
    "endTime": 1675502100000,
    "id": "E001",
    "livestream_endTime": 0,
    "prerecorded": true,
    "qa_startTime": null,
    "slug": "testcon_opening_remarks",
    "speakers": [
      {
        "email": "",
        "id": "9012",
        "matrix_id": "",
        "name": "John Teststone",
        "role": "speaker",
      },
      {
        "email": "",
        "id": "90367",
        "matrix_id": "",
        "name": "Evå 完牲 Exampleson",
        "role": "speaker",
      },
    ],
    "startTime": 1675501200000,
    "subtitle": "",
    "title": "Testcon01: Opening Remarks",
    "track": "Track01",
  },
]
`;
